src/main/java/uk/gov/hmcts/ccd/domain/service/stdapi/CallbackInvoker.java

```diff
@@ -228,4 +228,18 @@ private void deduceDataClassificationForNewFields(CaseType caseType, CaseDetails
         return jsonNode.flatMap(value -> value.isTextual() ? Optional.of(value.textValue()) : Optional.empty());
     }
 
+
+    void updateCallbackSateBasedOnPriority(final CallbackResponse callbackResponse) {
+        //it will work out the state value based on state priority.
+        //top level sate is the main priority. If there are defined states (top level and data state).
+        //the top level will take priority.
+        if ((callbackResponse.getState() == null || callbackResponse.getState().isEmpty())
+            && callbackResponse.getData() != null) {
+
+            final Optional<String> newCaseState = filterCaseState(callbackResponse.getData());
+            if (newCaseState.isPresent()) {
+                callbackResponse.setState(newCaseState.get());
+            }
+        }
+    }
 }
```

Comment:
I think it could be the `CallbackResponse` itself to expose a new method `getStateOverride` and most of this logic and also the one in `filterCaseState` could be properly encapsulated there. This will keep the CallbackInvoker simpler.